src/main/java/org/devnull/jedi/PowerDNSConnectionHandler.java

```diff
@@ -125,7 +126,14 @@ public void run()
 					log.debug("waiting to read request from socket");
 				}
 
-				requestLine = reader.readLine();
+				try {
+					requestLine = reader.readLine();
+				} catch (IOException e){
+					//expected, PowerDNS doesn't tell us when it's done using a connection, we need a timeout to clean up correctly.
+					log.debug("socket timed out, closing");
+					break;
```

Comment:
Yeah, we'd just get stuck reading from the socket again. The crux of the problem seems to be that the socket expects more data even after PowerDNS is done sending requests and the TCP connection is closed.